mava/utils/sebulba_utils.py

```diff
@@ -0,0 +1,186 @@
+
+    def should_stop(self) -> bool:
+        return self._stop
+
+    def stop(self) -> None:
+        self._stop = True
+
+
+class Pipeline(threading.Thread):
+    """
+    The `Pipeline` shards trajectories into `learner_devices`,
+    ensuring trajectories are consumed in the right order to avoid being off-policy
+    and limit the max number of samples in device memory at one time to avoid OOM issues.
+    """
+
+    def __init__(self, max_size: int, learner_devices: List[jax.Device], lifetime: ThreadLifetime):
+        """
+        Initializes the pipeline with a maximum size and the devices to shard trajectories across.
+
+        Args:
+            max_size: The maximum number of trajectories to keep in the pipeline.
+            learner_devices: The devices to shard trajectories across.
+        """
+        super().__init__(name="Pipeline")
+        self.learner_devices = learner_devices
+        self.tickets_queue: queue.Queue = queue.Queue()
+        self._queue: queue.Queue = queue.Queue(maxsize=max_size)
+        self.lifetime = lifetime
+
+    def run(self) -> None:
+        """This function ensures that trajectories on the queue are consumed in the right order. The
+        start_condition and end_condition are used to ensure that only 1 thread is processing an
+        item from the queue at one time, ensuring predictable memory usage.
+        """
+        while not self.lifetime.should_stop():
+            try:
+                start_condition, end_condition = self.tickets_queue.get(timeout=1)
+                with end_condition:
+                    with start_condition:
+                        start_condition.notify()
+                    end_condition.wait()
+            except queue.Empty:
+                continue
+
+    def put(self, traj: Sequence[PPOTransition], timestep: TimeStep, time_dict: Dict) -> None:
+        """Put a trajectory on the queue to be consumed by the learner."""
+        start_condition, end_condition = (threading.Condition(), threading.Condition())
+        with start_condition:
+            self.tickets_queue.put((start_condition, end_condition))
+            start_condition.wait()  # wait to be allowed to start
+
+        # [PPOTransition()] * rollout_len --> PPOTransition[done=(rollout_len, num_envs, num_agents)
+        sharded_traj = jax.tree.map(lambda *x: self.shard_split_playload(jnp.stack(x), 1), *traj)
+
+        # Timestep[(num_envs, num_agents, ...), ...] -->
+        # [(num_envs / num_learner_devices, num_agents, ...)] * num_learner_devices
+        sharded_timestep = jax.tree.map(self.shard_split_playload, timestep)
+
+        # If the queue gets full at any point we prioritize taking removing the oldest rollouts.
+        # This also prevents the pipeline from  stalling if the learner thread terminates
+        # with a full queue blocking the actors from placing items in it.
+        if self._queue.full():
+            self._queue.get()  # throw away the transition
```

Comment:
hello, I can definitely make these changes.
I have a few concerns tho:

- won't this lead to some threads dying in some configs (such as a large number of actors and a long rollout)
- Is ensuring reproducibility with non-blocking actors impossible?

And out of curiosity, why is discarding some experience considered cheating?

Reply:
Sorry for the late reply been a bit busy

1. Yes, but then I'd argue that you have made a bad config, especially if we make the queue timeout long enough - if an actor is waiting 90 seconds for a learn step then something is wrong with the learner
2. Not impossible but very hard
3. Because you usually report the reward on a graph where the x axis is steps and if we do steps but just ignore them (and don't report that those have been taken) then it's cheating. Think about if a simulator is really expensive you'd never want to throw away that data